File: src/main/java/com/medicare/backend/models/Invoice.java
package com.medicare.backend.models;

import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;

import java.time.LocalDateTime;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "invoices")
public class Invoice {
    @Id
    private String id;

    @Indexed(unique = true)
    private String invoiceId;

    @Indexed
    private String appointmentId;

    @Indexed
    private String patientId;

    private List<InvoiceItem> items;

    private Double total;

    @Indexed
    private String status = "pending"; // pending, paid

    private String paymentMethod; // Cash, Card, UPI, Insurance

    private Boolean hasInsurance = false;

    @CreatedDate
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class InvoiceItem {
        private String description;
        private Double amount;
    }
}

File: src/models/Invoice.ts
import mongoose, { Schema, Model } from 'mongoose';
import { IInvoice } from '../types';

const invoiceSchema = new Schema<IInvoice>(
    {
        invoiceId: {
            type: String,
            required: true,
            unique: true,
        },
        appointmentId: {
            type: String,
            required: true,
            ref: 'Appointment',
        },
        patientId: {
            type: String,
            required: true,
            ref: 'Patient',
        },
        items: [
            {
                description: { type: String, required: true },
                amount: { type: Number, required: true },
            },
        ],
        total: {
            type: Number,
            required: true,
        },
        status: {
            type: String,
            enum: ['pending', 'paid'],
            default: 'pending',
        },
        paymentMethod: {
            type: String,
            enum: ['Cash', 'Card', 'UPI', 'Insurance'],
        },
        hasInsurance: {
            type: Boolean,
            default: false,
        },
    },
    {
        timestamps: true,
    }
);

// Create indexes
invoiceSchema.index({ invoiceId: 1 });
invoiceSchema.index({ appointmentId: 1 });
invoiceSchema.index({ patientId: 1 });
invoiceSchema.index({ status: 1 });

export const Invoice: Model<IInvoice> = mongoose.model<IInvoice>('Invoice', invoiceSchema);
